### ISSUE TITLE
OSX dispatch issue - socket is not connected

### ISSUE THREAD

**Issue by jimmyyih** [greenplum-db/gpdb#1164]
Compiling and running the latest commit 78b8726c47e9ffec0e3118467e16f2f953b9316f on my OSX Yosemite box and on another OSX El Capitan box, I frequently run into some dispatch errors while running pg_regress tests on a gpdemo cluster (master, 3 primaries, 3 mirrors).

Master pg_log:
```
2016-09-28 14:55:30.861136 PDT,"jyih","regression",p92385,th2070401792,"[local]",,2016-09-28 14:55:30 PDT,2248,con90,cmd33,seg-1,,dx1545,x2248,sx1,"ERROR","58M01","failed to acquire resources on one or more segments","could not send data to server: Socket is not connected
could not send startup packet: Socket is not connected
 (seg2 127.0.0.1:25434)",,,,,"SELECT '' AS fifteen, r1.*, r2.*
   FROM RELTIME_TBL r1, RELTIME_TBL r2
   WHERE r1.f1 > r2.f1
   ORDER BY r1.f1, r2.f1;",0,,"cdbgang_async.c",182,

..... (skip some uninteresting logs) .....

2016-09-28 14:55:33.492731 PDT,"jyih","regression",p92510,th2070401792,"[local]",,2016-09-28 14:55:33 PDT,2860,con97,,seg-1,,dx2157,x2860,sx1,"LOG","00000","The previous session was reset because its gang was disconnected (session id = 92). The new session id = 97",,,,,,"CREATE TABLE ABSTIME_HOROLOGY_TBL (f1 abstime);",0,,"cdbgang.c",1504,
2016-09-28 14:55:33.492794 PDT,"jyih","regression",p92510,th2070401792,"[local]",,2016-09-28 14:55:33 PDT,2860,con97,,seg-1,,dx2157,x2860,sx1,"ERROR","58M01","failed to acquire resources on one or more segments","could not send data to server: Socket is not connected
could not send startup packet: Socket is not connected
 (seg2 127.0.0.1:25434)",,,,,"CREATE TABLE ABSTIME_HOROLOGY_TBL (f1 abstime);",0,,"cdbgang_async.c",182,
```

seg2 pg_log:
```
2016-09-28 14:55:30.823841 PDT,,,p92417,th2070401792,"127.0.0.1","59665",2016-09-28 14:55:30 PDT,0,,,seg-1,,,,,"LOG","00000","query finish request to process 92398",,,,,,,0,,"postmaster.c",3742,
2016-09-28 14:55:31.110911 PDT,,,p92420,th2070401792,"127.0.0.1","59670",2016-09-28 14:55:30 PDT,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2643,
2016-09-28 14:55:34.255942 PDT,,,p92519,th2070401792,"127.0.0.1","59692",2016-09-28 14:55:33 PDT,0,,,seg-1,,,,,"LOG","08P01","incomplete startup packet",,,,,,,0,,"postmaster.c",2643,
2016-09-28 14:55:35.993784 PDT,,,p90697,th2070401792,,,,0,con1,,seg-1,,,,,"LOG","00000","activeWeight underflow!",,,,,,,0,,"backoff.c",783,
```

It doesn't seem to error on CentOS though.  Any Greenplum developers hacking on OSX running into this issue or have fixed it via environment changes?

@eurekaka, @pengzhout, @xiong-gang: Any suggestions since it looks like you guys are touching the area?

**Comment by danielgustafsson** [comment]
I have seen the same issues on OS X El Capitan. I think it started sometime when I was away since I didn't see it before being on leave but had it happen immediately when I came back (so between July 22 and August 22). Haven't had time to dig in to the details and figure out why yet.

**Comment by xiong-gang** [comment]
We can reproduce on OS X, we are looking for the root cause.
Thanks for reporting it.

**Comment by eurekaka** [comment]
Here is my finding:

QD calls an asynchronous connect() to make a TCP connection to segment 2,
after that, QD calls poll() on the socket to wait for the socket being
ready for writing; after poll() tells that the socket is ready for writing,
QD calls send() to send libpq startup packet to the QE started on segment
2, but send() returns -1 and the errno is ENOTCONN, which means the socket
is not connected, however, by using netstat or lsof, we see that the
connection is in "ESTABLISHED" state for both QD and QE side, and by using
tcpdump, we notice that there is no 'FIN' packet from both side to close
the connection. That is to say, no one has closed the socket and the TCP
connection is established as seen by OSX while the send() call would fail.

Then I dive into the packet exchange of this problematic TCP connection,
here is the tcpdump result:

18:17:23.834904 IP imac.50968 > imac.25434: Flags [SEW], seq
2606128632, win 65535, options [mss 16344,nop,wscale 5,nop,nop,TS val
120992810 ecr 0,sackOK,eol], length 0
18:17:23.835012 IP imac.25434 > imac.50968: Flags [S.E], seq
4038226809, ack 2606128633, win 65535, options [mss 16344,nop,wscale
5,nop,nop,TS val 120992810 ecr 120992810,sackOK,eol], length 0
18:17:23.835035 IP imac.50968 > imac.25434: Flags [.], ack 1, win
12759, options [nop,nop,TS val 120992810 ecr 120992810], length 0
18:17:23.835050 IP imac.25434 > imac.50968: Flags [.], ack 1, win
12759, options [nop,nop,TS val 120992810 ecr 120992810], length 0

​
and here is the dump result of a normal connection from QD to QE:

18:17:23.826102 IP imac.50965 > imac.25434: Flags [S], seq 3793144569,
win 65535, options [mss 16344,nop,wscale 5,nop,nop,TS val 120992802
ecr 0,sackOK,eol], length 0
18:17:23.826139 IP imac.25434 > imac.50965: Flags [S.], seq
3486547735, ack 3793144570, win 65535, options [mss 16344,nop,wscale
5,nop,nop,TS val 120992802 ecr 120992802,sackOK,eol], length 0
18:17:23.826144 IP imac.50965 > imac.25434: Flags [.], ack 1, win
12759, options [nop,nop,TS val 120992802 ecr 120992802], length 0
18:17:23.826150 IP imac.25434 > imac.50965: Flags [.], ack 1, win
12759, options [nop,nop,TS val 120992802 ecr 120992802], length 0
18:17:23.826206 IP imac.50965 > imac.25434: Flags [.], seq 1:16333,
ack 1, win 12759, options [nop,nop,TS val 120992802 ecr 120992802],
length 16332*CRITICAL PACKET* ==> 18:17:23.826208 IP imac.50965 >
imac.25434: Flags [P.], seq 16333:16385, ack 1, win 12759, options
[nop,nop,TS val 120992802 ecr 120992802], length 52
18:17:23.826212 IP imac.50965 > imac.25434: Flags [P.], seq
16385:21964, ack 1, win 12759, options [nop,nop,TS val 120992802 ecr
120992802], length 5579
18:17:23.826215 IP imac.25434 > imac.50965: Flags [.], ack 16385, win
12247, options [nop,nop,TS val 120992802 ecr 120992802], length 0
18:17:23.826218 IP imac.25434 > imac.50965: Flags [.], ack 21964, win
12073, options [nop,nop,TS val 120992802 ecr 120992802], length 0
18:17:23.833886 IP imac.25434 > imac.50965: Flags [P.], seq 1:492, ack
21964, win 12073, options [nop,nop,TS val 120992809 ecr 120992802],
length 491

​
we can see that, in the normal connection, there are 5 packet exchanges to
establish a connection(it is strange though, because for the 3-way
handshake protocol, there should be only 3 packet exchanges, I googled
around why OSX is using this kind of "5-way handshake" but found nothing,
anyone knows about this?), but in the problematic connection, there are
only 4 packet exchanges, and the last ACK from QD to QE is missed.

Here is the dump result on CentOS 7, and we can see there are only 3 packet
exchanges to establish a TCP connection:

19:05:42.733994 IP localhost.54524 > localhost.25434: Flags [S], seq
2645036561, win 43690, options [mss 65495,sackOK,TS val 832575014 ecr
0,nop,wscale 7], length 0
19:05:42.734023 IP localhost.25434 > localhost.54524: Flags [S.], seq
100572218, ack 2645036562, win 43690, options [mss 65495,sackOK,TS val
832575014 ecr 832575014,nop,wscale 7], length 0
19:05:42.734046 IP localhost.54524 > localhost.25434: Flags [.], ack
1, win 342, options [nop,nop,TS val 832575014 ecr 832575014], length 0
19:05:42.734466 IP localhost.54524 > localhost.25434: Flags [P.], seq
1:16385, ack 1, win 342, options [nop,nop,TS val 832575015 ecr
832575014], length 16384
19:05:42.734487 IP localhost.25434 > localhost.54524: Flags [.], ack
16385, win 957, options [nop,nop,TS val 832575015 ecr 832575015],
length 0

​

Therefore, from my personal point of view, this bug is not in GPDB, it
should be a bug of OSX, because the handshake was actually not completed,
but OS believed the connection was established, and poll() believed the
socket was ready for writing then, which indeed was not and hence send()
call failed.

The explanation of QE's error log is: QD cannot send startup packet to QE,
and report ERROR then close this socket, while QE is waiting for the
startup packet from QD on pq_recvmsg -> recv, then it detect the close of
the connection from QD, and returns 0 from recv, then function
ProcessStartupPacket believes it receives the EOF and report the error
"incomplete startup packet".

One more thing worth mention is that this bug only happens for asynchronous
dispatcher on OSX, and it is not reproducible when using threaded
dispatcher.

Cheers,
Kenan

**Comment by jimmyyih** [comment]
Thanks Kenan for the detailed RCA!  It is very insightful.  For now, I will use the GUC `gp_connections_per_thread` (setting to 64) to get around the asynchronous dispatcher issues on OSX.

**Comment by danielgustafsson** [comment]
Even using the threaded dispatcher I see errors, have others observed that as well or am I just doing something wrong?

**Comment by ashwinstar** [comment]
With threaded dispatcher atleast I am not hitting the error now which used
to happen extremely frequently earlier. Like even basic ICG tests int64 and
all used fail, not seeing errors with threaded dispatcher.

But its certainly annoying to remember one more tweak to work on OSX.

**Comment by pengzhout** [comment]
we found a potential defect that cause this issue, will draft a PR to fix this.

**Comment by ashwinstar** [comment]
Thank You that would be very helpful.

**Comment by asimrp** [comment]
Hey @eurekaka, thank you for digging into the details of packet exchange.  I found it very helpful.  The mystery of "5-way handshake" can be explained with TCP window scaling, as defined in RFC 7323: https://tools.ietf.org/html/rfc7323

Notice the window size being exchanged in your packet trace.  Let me quote the packets again for convenience: 
```
18:17:23.826102 IP imac.50965 > imac.25434: Flags [S], seq 3793144569, win 65535, options [mss 16344,nop,wscale 5,nop,nop,TS val 120992802 ecr 0,sackOK,eol], length 0

18:17:23.826139 IP imac.25434 > imac.50965: Flags [S.], seq 3486547735, ack 3793144570, win 65535, options [mss 16344,nop,wscale 5,nop,nop,TS val 120992802 ecr 120992802,sackOK,eol], length 0

18:17:23.826144 IP imac.50965 > imac.25434: Flags [.], ack 1, win 12759, options [nop,nop,TS val 120992802 ecr 120992802], length 0

18:17:23.826150 IP imac.25434 > imac.50965: Flags [.], ack 1, win 12759, options [nop,nop,TS val 120992802 ecr 120992802], length 0
```
Window size in the first two (SYN and SYN-ACK) packets is `win 65535` with window scale of 5 (`wscale 5`).  Whereas the actual window size, `win 12759` is negotiated in the following two ack packets.

I tried capturing packets after disabling window scaling on my OSX laptop and it turned out to be a regular three-way handshake.

```
# sysctl net.inet.tcp.rfc1323
net.inet.tcp.rfc1323: 1
# sysctl -w net.inet.tcp.rfc1323=0
net.inet.tcp.rfc1323: 1 -> 0
```
Capture all tcp packets on loopback interface for an ssh connection.  Note that the fourth packet has data length greater than 0, implying that connection was established after the first three packets.

```
# tcpdump -S -c 10 -i lo0 'tcp port 22'
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on lo0, link-type NULL (BSD loopback), capture size 65535 bytes
16:17:15.165884 IP6 localhost.50881 > localhost.ssh: Flags [S], seq 995642791, win 65535, options [mss 16324,sackOK,eol], length 0
16:17:15.165939 IP6 localhost.ssh > localhost.50881: Flags [S.], seq 422264072, ack 995642792, win 65535, options [mss 16324,sackOK,eol], length 0
16:17:15.165954 IP6 localhost.50881 > localhost.ssh: Flags [.], ack 422264073, win 65535, length 0
16:17:15.166434 IP6 localhost.50881 > localhost.ssh: Flags [P.], seq 995642792:995642813, ack 422264073, win 65535, length 21
```

**Comment by pengzhout** [comment]
fixed , patch is f626e74dda981f7a71e88f552284c60b11aac856

**Comment by eurekaka** [comment]
Thanks @asimrp for the clarification, it helps, and thanks @pengzhout for the fix, it is awesome.
